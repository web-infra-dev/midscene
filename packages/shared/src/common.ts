import fs from 'node:fs';
import path from 'node:path';

// Define locally for now to avoid import issues
export const isNodeEnv =
  typeof process !== 'undefined' &&
  process.versions != null &&
  process.versions.node != null;

/**
 * Get the path to the midscene_run directory or a subdirectory within it.
 * Creates the directory if it doesn't exist.
 *
 * @param subdir - Optional subdirectory name (e.g., 'log', 'report')
 * @returns The absolute path to the requested directory
 */
export const getMidsceneRunLogPath = (): string => {
  const basePath = path.join(process.cwd(), 'midscene_run');

  // Create a base directory
  if (!fs.existsSync(basePath)) {
    fs.mkdirSync(basePath, { recursive: true });
  }

  // Create a log directory
  const logPath = path.join(basePath, 'log');
  if (!fs.existsSync(logPath)) {
    fs.mkdirSync(logPath, { recursive: true });
  }

  return logPath;
};

export const logDir = isNodeEnv ? getMidsceneRunLogPath() : '';
