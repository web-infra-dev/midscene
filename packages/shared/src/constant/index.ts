export const TEXT_SIZE_THRESHOLD = 9;

export const TEXT_MAX_SIZE = 40;

export enum NodeType {
  INPUT = 'INPUT Node',
  BUTTON = 'BUTTON Node',
  IMG = 'IMG Node',
  TEXT = 'TEXT Node',
}
