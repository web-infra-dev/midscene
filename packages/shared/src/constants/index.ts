export const TEXT_SIZE_THRESHOLD = 9;

export const TEXT_MAX_SIZE = 40;

export const CONTAINER_MINI_HEIGHT = 3;
export const CONTAINER_MINI_WIDTH = 3;

export enum NodeType {
  CONTAINER = 'CONTAINER Node',
  FORM_ITEM = 'FORM_ITEM Node',
  BUTTON = 'BUTTON Node',
  IMG = 'IMG Node',
  TEXT = 'TEXT Node',
}
