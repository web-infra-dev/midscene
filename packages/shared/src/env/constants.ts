import {
  ANTHROPIC_API_KEY,
  AZURE_OPENAI_API_VERSION,
  AZURE_OPENAI_DEPLOYMENT,
  AZURE_OPENAI_ENDPOINT,
  AZURE_OPENAI_KEY,
  MIDSCENE_ANTHROPIC_API_KEY,
  MIDSCENE_AZURE_OPENAI_API_VERSION,
  MIDSCENE_AZURE_OPENAI_DEPLOYMENT,
  MIDSCENE_AZURE_OPENAI_ENDPOINT,
  MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_AZURE_OPENAI_KEY,
  MIDSCENE_AZURE_OPENAI_SCOPE,
  MIDSCENE_GROUNDING_ANTHROPIC_API_KEY,
  MIDSCENE_GROUNDING_AZURE_OPENAI_API_VERSION,
  MIDSCENE_GROUNDING_AZURE_OPENAI_DEPLOYMENT,
  MIDSCENE_GROUNDING_AZURE_OPENAI_ENDPOINT,
  MIDSCENE_GROUNDING_AZURE_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_GROUNDING_AZURE_OPENAI_KEY,
  MIDSCENE_GROUNDING_AZURE_OPENAI_SCOPE,
  MIDSCENE_GROUNDING_MODEL_NAME,
  MIDSCENE_GROUNDING_OPENAI_API_KEY,
  MIDSCENE_GROUNDING_OPENAI_BASE_URL,
  MIDSCENE_GROUNDING_OPENAI_HTTP_PROXY,
  MIDSCENE_GROUNDING_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_GROUNDING_OPENAI_SOCKS_PROXY,
  MIDSCENE_GROUNDING_OPENAI_USE_AZURE,
  MIDSCENE_GROUNDING_USE_ANTHROPIC_SDK,
  MIDSCENE_GROUNDING_USE_AZURE_OPENAI,
  MIDSCENE_GROUNDING_VL_MODE,
  MIDSCENE_MODEL_NAME,
  MIDSCENE_OPENAI_API_KEY,
  MIDSCENE_OPENAI_BASE_URL,
  MIDSCENE_OPENAI_HTTP_PROXY,
  MIDSCENE_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_OPENAI_SOCKS_PROXY,
  MIDSCENE_OPENAI_USE_AZURE,
  MIDSCENE_PLANNING_ANTHROPIC_API_KEY,
  MIDSCENE_PLANNING_AZURE_OPENAI_API_VERSION,
  MIDSCENE_PLANNING_AZURE_OPENAI_DEPLOYMENT,
  MIDSCENE_PLANNING_AZURE_OPENAI_ENDPOINT,
  MIDSCENE_PLANNING_AZURE_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_PLANNING_AZURE_OPENAI_KEY,
  MIDSCENE_PLANNING_AZURE_OPENAI_SCOPE,
  MIDSCENE_PLANNING_MODEL_NAME,
  MIDSCENE_PLANNING_OPENAI_API_KEY,
  MIDSCENE_PLANNING_OPENAI_BASE_URL,
  MIDSCENE_PLANNING_OPENAI_HTTP_PROXY,
  MIDSCENE_PLANNING_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_PLANNING_OPENAI_SOCKS_PROXY,
  MIDSCENE_PLANNING_OPENAI_USE_AZURE,
  MIDSCENE_PLANNING_USE_ANTHROPIC_SDK,
  MIDSCENE_PLANNING_USE_AZURE_OPENAI,
  MIDSCENE_PLANNING_VL_MODE,
  MIDSCENE_USE_ANTHROPIC_SDK,
  MIDSCENE_USE_AZURE_OPENAI,
  MIDSCENE_VL_MODE,
  MIDSCENE_VQA_ANTHROPIC_API_KEY,
  MIDSCENE_VQA_AZURE_OPENAI_API_VERSION,
  MIDSCENE_VQA_AZURE_OPENAI_DEPLOYMENT,
  MIDSCENE_VQA_AZURE_OPENAI_ENDPOINT,
  MIDSCENE_VQA_AZURE_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_VQA_AZURE_OPENAI_KEY,
  MIDSCENE_VQA_AZURE_OPENAI_SCOPE,
  // VQA
  MIDSCENE_VQA_MODEL_NAME,
  MIDSCENE_VQA_OPENAI_API_KEY,
  MIDSCENE_VQA_OPENAI_BASE_URL,
  MIDSCENE_VQA_OPENAI_HTTP_PROXY,
  MIDSCENE_VQA_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_VQA_OPENAI_SOCKS_PROXY,
  MIDSCENE_VQA_OPENAI_USE_AZURE,
  MIDSCENE_VQA_USE_ANTHROPIC_SDK,
  MIDSCENE_VQA_USE_AZURE_OPENAI,
  MIDSCENE_VQA_VL_MODE,
  OPENAI_API_KEY,
  OPENAI_BASE_URL,
  OPENAI_USE_AZURE,
} from './types';

interface IModelConfigKeys {
  modelName: string;
  /**
   * proxy
   */
  socksProxy: string;
  httpProxy: string;
  /**
   * OpenAI
   */
  openaiBaseURL: string;
  openaiApiKey: string;
  openaiExtraConfig: string;
  /**
   * Azure
   */
  openaiUseAzureDeprecated: string;
  useAzureOpenai: string;
  azureOpenaiScope: string;
  azureOpenaiKey: string;
  azureOpenaiEndpoint: string;
  azureOpenaiApiVersion: string;
  azureOpenaiDeployment: string;
  azureExtraConfig: string;
  /**
   * Anthropic
   */
  useAnthropicSdk: string;
  anthropicApiKey: string;
  /**
   * Extra
   */
  vlMode: string;
}

export const VQA_MODEL_CONFIG_KEYS: IModelConfigKeys = {
  modelName: MIDSCENE_VQA_MODEL_NAME,
  /**
   * proxy
   */
  socksProxy: MIDSCENE_VQA_OPENAI_SOCKS_PROXY,
  httpProxy: MIDSCENE_VQA_OPENAI_HTTP_PROXY,
  /**
   * OpenAI
   */
  openaiBaseURL: MIDSCENE_VQA_OPENAI_BASE_URL,
  openaiApiKey: MIDSCENE_VQA_OPENAI_API_KEY,
  openaiExtraConfig: MIDSCENE_VQA_OPENAI_INIT_CONFIG_JSON,
  /**
   * Azure
   */
  openaiUseAzureDeprecated: MIDSCENE_VQA_OPENAI_USE_AZURE,
  useAzureOpenai: MIDSCENE_VQA_USE_AZURE_OPENAI,
  azureOpenaiScope: MIDSCENE_VQA_AZURE_OPENAI_SCOPE,
  azureOpenaiKey: MIDSCENE_VQA_AZURE_OPENAI_KEY,
  azureOpenaiEndpoint: MIDSCENE_VQA_AZURE_OPENAI_ENDPOINT,
  azureOpenaiApiVersion: MIDSCENE_VQA_AZURE_OPENAI_API_VERSION,
  azureOpenaiDeployment: MIDSCENE_VQA_AZURE_OPENAI_DEPLOYMENT,
  azureExtraConfig: MIDSCENE_VQA_AZURE_OPENAI_INIT_CONFIG_JSON,
  /**
   * Anthropic
   */
  useAnthropicSdk: MIDSCENE_VQA_USE_ANTHROPIC_SDK,
  anthropicApiKey: MIDSCENE_VQA_ANTHROPIC_API_KEY,
  /**
   * Extra
   */
  vlMode: MIDSCENE_VQA_VL_MODE,
} as const;

export const GROUNDING_MODEL_CONFIG_KEYS: IModelConfigKeys = {
  modelName: MIDSCENE_GROUNDING_MODEL_NAME,
  /**
   * proxy
   */
  socksProxy: MIDSCENE_GROUNDING_OPENAI_SOCKS_PROXY,
  httpProxy: MIDSCENE_GROUNDING_OPENAI_HTTP_PROXY,
  /**
   * OpenAI
   */
  openaiBaseURL: MIDSCENE_GROUNDING_OPENAI_BASE_URL,
  openaiApiKey: MIDSCENE_GROUNDING_OPENAI_API_KEY,
  openaiExtraConfig: MIDSCENE_GROUNDING_OPENAI_INIT_CONFIG_JSON,
  /**
   * Azure
   */
  openaiUseAzureDeprecated: MIDSCENE_GROUNDING_OPENAI_USE_AZURE,
  useAzureOpenai: MIDSCENE_GROUNDING_USE_AZURE_OPENAI,
  azureOpenaiScope: MIDSCENE_GROUNDING_AZURE_OPENAI_SCOPE,
  azureOpenaiKey: MIDSCENE_GROUNDING_AZURE_OPENAI_KEY,
  azureOpenaiEndpoint: MIDSCENE_GROUNDING_AZURE_OPENAI_ENDPOINT,
  azureOpenaiApiVersion: MIDSCENE_GROUNDING_AZURE_OPENAI_API_VERSION,
  azureOpenaiDeployment: MIDSCENE_GROUNDING_AZURE_OPENAI_DEPLOYMENT,
  azureExtraConfig: MIDSCENE_GROUNDING_AZURE_OPENAI_INIT_CONFIG_JSON,
  /**
   * Anthropic
   */
  useAnthropicSdk: MIDSCENE_GROUNDING_USE_ANTHROPIC_SDK,
  anthropicApiKey: MIDSCENE_GROUNDING_ANTHROPIC_API_KEY,
  /**
   * Extra
   */
  vlMode: MIDSCENE_GROUNDING_VL_MODE,
} as const;

export const PLANNING_MODEL_CONFIG_KEYS: IModelConfigKeys = {
  modelName: MIDSCENE_PLANNING_MODEL_NAME,
  /**
   * proxy
   */
  socksProxy: MIDSCENE_PLANNING_OPENAI_SOCKS_PROXY,
  httpProxy: MIDSCENE_PLANNING_OPENAI_HTTP_PROXY,
  /**
   * OpenAI
   */
  openaiBaseURL: MIDSCENE_PLANNING_OPENAI_BASE_URL,
  openaiApiKey: MIDSCENE_PLANNING_OPENAI_API_KEY,
  openaiExtraConfig: MIDSCENE_PLANNING_OPENAI_INIT_CONFIG_JSON,
  /**
   * Azure
   */
  openaiUseAzureDeprecated: MIDSCENE_PLANNING_OPENAI_USE_AZURE,
  useAzureOpenai: MIDSCENE_PLANNING_USE_AZURE_OPENAI,
  azureOpenaiScope: MIDSCENE_PLANNING_AZURE_OPENAI_SCOPE,
  azureOpenaiKey: MIDSCENE_PLANNING_AZURE_OPENAI_KEY,
  azureOpenaiEndpoint: MIDSCENE_PLANNING_AZURE_OPENAI_ENDPOINT,
  azureOpenaiApiVersion: MIDSCENE_PLANNING_AZURE_OPENAI_API_VERSION,
  azureOpenaiDeployment: MIDSCENE_PLANNING_AZURE_OPENAI_DEPLOYMENT,
  azureExtraConfig: MIDSCENE_PLANNING_AZURE_OPENAI_INIT_CONFIG_JSON,
  /**
   * Anthropic
   */
  useAnthropicSdk: MIDSCENE_PLANNING_USE_ANTHROPIC_SDK,
  anthropicApiKey: MIDSCENE_PLANNING_ANTHROPIC_API_KEY,
  /**
   * Extra
   */
  vlMode: MIDSCENE_PLANNING_VL_MODE,
} as const;

// modelConfig return default
export const DEFAULT_MODEL_CONFIG_KEYS: IModelConfigKeys = {
  modelName: MIDSCENE_MODEL_NAME,
  /**
   * proxy
   */
  socksProxy: MIDSCENE_OPENAI_SOCKS_PROXY,
  httpProxy: MIDSCENE_OPENAI_HTTP_PROXY,
  /**
   * OpenAI
   */
  openaiBaseURL: MIDSCENE_OPENAI_BASE_URL,
  openaiApiKey: MIDSCENE_OPENAI_API_KEY,
  openaiExtraConfig: MIDSCENE_OPENAI_INIT_CONFIG_JSON,
  /**
   * Azure
   */
  openaiUseAzureDeprecated: MIDSCENE_OPENAI_USE_AZURE,
  useAzureOpenai: MIDSCENE_USE_AZURE_OPENAI,
  azureOpenaiScope: MIDSCENE_AZURE_OPENAI_SCOPE,
  azureOpenaiKey: MIDSCENE_AZURE_OPENAI_KEY,
  azureOpenaiEndpoint: MIDSCENE_AZURE_OPENAI_ENDPOINT,
  azureOpenaiApiVersion: MIDSCENE_AZURE_OPENAI_API_VERSION,
  azureOpenaiDeployment: MIDSCENE_AZURE_OPENAI_DEPLOYMENT,
  azureExtraConfig: MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON,
  /**
   * Anthropic
   */
  useAnthropicSdk: MIDSCENE_USE_ANTHROPIC_SDK,
  anthropicApiKey: MIDSCENE_ANTHROPIC_API_KEY,
  /**
   * Extra
   */
  vlMode: MIDSCENE_VL_MODE,
} as const;

// read from process.env
export const DEFAULT_MODEL_CONFIG_KEYS_LEGACY: IModelConfigKeys = {
  modelName: MIDSCENE_MODEL_NAME,
  /**
   * proxy
   */
  socksProxy: MIDSCENE_OPENAI_SOCKS_PROXY,
  httpProxy: MIDSCENE_OPENAI_HTTP_PROXY,
  /**
   * OpenAI
   */
  openaiBaseURL: OPENAI_BASE_URL,
  openaiApiKey: OPENAI_API_KEY,
  openaiExtraConfig: MIDSCENE_OPENAI_INIT_CONFIG_JSON,
  /**
   * Azure
   */
  openaiUseAzureDeprecated: OPENAI_USE_AZURE,
  useAzureOpenai: MIDSCENE_USE_AZURE_OPENAI,
  azureOpenaiScope: MIDSCENE_AZURE_OPENAI_SCOPE,
  azureOpenaiKey: AZURE_OPENAI_KEY,
  azureOpenaiEndpoint: AZURE_OPENAI_ENDPOINT,
  azureOpenaiApiVersion: AZURE_OPENAI_API_VERSION,
  azureOpenaiDeployment: AZURE_OPENAI_DEPLOYMENT,
  azureExtraConfig: MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON,
  /**
   * Anthropic
   */
  useAnthropicSdk: MIDSCENE_USE_ANTHROPIC_SDK,
  anthropicApiKey: ANTHROPIC_API_KEY,
  /**
   * Extra
   */
  vlMode: 'DEFAULT_MODEL_CONFIG_KEYS has no vlMode key',
} as const;
