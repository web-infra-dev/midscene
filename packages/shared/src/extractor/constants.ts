export {
  NodeType,
  TEXT_MAX_SIZE,
  TEXT_SIZE_THRESHOLD,
} from '../constants/index';
