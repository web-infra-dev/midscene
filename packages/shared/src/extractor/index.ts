import type { NodeType } from '../constants/index';

export interface ElementInfo {
  id: string;
  indexId: number;
  nodeHashId: string;
  locator: string;
  attributes: {
    nodeType: NodeType;
    [key: string]: string;
  };
  nodeType: NodeType;
  content: string;
  rect: { left: number; top: number; width: number; height: number };
  center: [number, number];
}

export interface ElementNode {
  node: ElementInfo | null;
  children: ElementNode[];
}

export { descriptionOfTree, traverseTree, treeToList } from './tree';

export { extractTextWithPosition as webExtractTextWithPosition } from './web-extractor';

export { extractTextWithPosition as clientExtractTextWithPosition } from './client-extractor';

export { extractTreeNode as webExtractNodeTree } from './web-extractor';

export { extractTreeNodeAsString as webExtractNodeTreeAsString } from './web-extractor';
