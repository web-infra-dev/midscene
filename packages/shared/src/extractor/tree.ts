import type { BaseElement, ElementTreeNode } from '../types';

export function truncateText(
  text: string | number | object | undefined,
  maxLength = 150,
) {
  if (typeof text === 'undefined') {
    return '';
  }

  if (typeof text === 'object') {
    text = JSON.stringify(text);
  }

  if (typeof text === 'number') {
    return text.toString();
  }

  if (typeof text === 'string' && text.length > maxLength) {
    return `${text.slice(0, maxLength)}...`;
  }

  if (typeof text === 'string') {
    return text.trim();
  }

  return '';
}

export function trimAttributes(
  attributes: Record<string, any>,
  truncateTextLength?: number,
) {
  const tailorAttributes = Object.keys(attributes).reduce(
    (res, currentKey: string) => {
      const attributeVal = (attributes as any)[currentKey];
      if (
        currentKey === 'style' ||
        currentKey === 'src' ||
        currentKey === 'htmlTagName' ||
        currentKey === 'nodeType'
      ) {
        return res;
      }

      res[currentKey] = truncateText(attributeVal, truncateTextLength);
      return res;
    },
    {} as BaseElement['attributes'],
  );
  return tailorAttributes;
}

const nodeSizeThreshold = 4;
export function descriptionOfTree<
  ElementType extends BaseElement = BaseElement,
>(
  tree: ElementTreeNode<ElementType>,
  truncateTextLength?: number,
  filterNonTextContent = false,
) {
  const attributesString = (kv: Record<string, any>) => {
    return Object.entries(kv)
      .map(
        ([key, value]) => `${key}="${truncateText(value, truncateTextLength)}"`,
      )
      .join(' ');
  };

  function buildContentTree(
    node: ElementTreeNode<ElementType>,
    indent = 0,
  ): string {
    let before = '';
    let contentWithIndent = '';
    let after = '';
    let emptyNode = true;
    const indentStr = '  '.repeat(indent);

    let children = '';
    for (let i = 0; i < (node.children || []).length; i++) {
      const childContent = buildContentTree(node.children[i], indent + 1);
      if (childContent) {
        children += `\n${childContent}`;
      }
    }

    if (
      node.node &&
      node.node.rect.width > nodeSizeThreshold &&
      node.node.rect.height > nodeSizeThreshold &&
      (!filterNonTextContent || (filterNonTextContent && node.node.content))
    ) {
      emptyNode = false;
      let nodeTypeString: string;
      if (node.node.attributes?.htmlTagName) {
        nodeTypeString = node.node.attributes.htmlTagName.replace(/[<>]/g, '');
      } else {
        nodeTypeString = node.node.attributes.nodeType
          .replace(/\sNode$/, '')
          .toLowerCase();
      }
      const markerId = node.node.indexId;
      const markerIdString = markerId ? `markerId="${markerId}"` : '';
      const rectAttribute = node.node.rect
        ? {
            left: node.node.rect.left,
            top: node.node.rect.top,
            width: node.node.rect.width,
            height: node.node.rect.height,
          }
        : {};
      before = `<${nodeTypeString} id="${node.node.id}" ${markerIdString} ${attributesString(trimAttributes(node.node.attributes || {}, truncateTextLength))} ${attributesString(rectAttribute)}>`;
      const content = truncateText(node.node.content, truncateTextLength);
      contentWithIndent = content ? `\n${indentStr}  ${content}` : '';
      after = `</${nodeTypeString}>`;
    } else if (!filterNonTextContent) {
      if (!children.trim().startsWith('<>')) {
        before = '<>';
        contentWithIndent = '';
        after = '</>';
      }
    }

    if (emptyNode && !children.trim()) {
      return '';
    }

    const result = `${indentStr}${before}${contentWithIndent}${children}\n${indentStr}${after}`;
    if (result.trim()) {
      return result;
    }
    return '';
  }

  const result = buildContentTree(tree);
  return result.replace(/^\s*\n/gm, '');
}

export function treeToList<T extends BaseElement>(
  tree: ElementTreeNode<T>,
): T[] {
  const result: T[] = [];
  function dfs(node: ElementTreeNode<T>) {
    if (node.node) {
      result.push(node.node);
    }
    for (const child of node.children) {
      dfs(child);
    }
  }
  dfs(tree);
  return result;
}

export function traverseTree<
  T extends BaseElement,
  ReturnNodeType extends BaseElement,
>(
  tree: ElementTreeNode<T>,
  onNode: (node: T) => ReturnNodeType,
): ElementTreeNode<ReturnNodeType> {
  function dfs(node: ElementTreeNode<T>) {
    if (node.node) {
      node.node = onNode(node.node) as any;
    }
    for (const child of node.children) {
      dfs(child);
    }
  }
  dfs(tree);
  return tree as any;
}
