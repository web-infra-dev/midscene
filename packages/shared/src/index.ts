export {
  createCopyStaticPlugin,
  createPlaygroundCopyPlugin,
} from './build/copy-static';

export default {};
