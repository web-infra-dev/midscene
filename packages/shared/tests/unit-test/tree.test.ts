import { describe, expect, it } from 'vitest';
import { NodeType } from '../../src/constants';
import { descriptionOfTree } from '../../src/extractor/tree';

describe('utils', () => {
  const tree = {
    node: {
      attributes: {
        nodeType: NodeType.CONTAINER,
      },
      id: '1',
      indexId: 19,
      rect: {
        x: 0,
        y: 0,
        width: 100,
        height: 100,
        left: 0,
        top: 0,
      },
      center: [50, 50] as [number, number],
      content:
        'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
    },
    children: [
      {
        node: {
          attributes: {
            nodeType: NodeType.TEXT,
            ariaLabel:
              'image description, it could be a long text, very loooooooooooooooooooooooooooooooooooooooooong',
          },
          id: '2',
          indexId: 999,
          rect: {
            x: 0,
            y: 0,
            width: 100,
            height: 100,
            left: 0,
            top: 0,
          },
          center: [50, 50] as [number, number],
          content: 'world',
        },
        children: [],
      },
      {
        node: {
          attributes: {
            nodeType: NodeType.IMG,
            style: 'width: 100px; height: 100px;',
            src: 'https://example.com/image.jpg',
            htmlTagName: '<img>',
            ariaLabel: 'image description',
            storyContent:
              'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
          },
          id: '3',
          indexId: 20,
          rect: {
            x: 0,
            y: 0,
            width: 100,
            height: 100,
            left: 0,
            top: 0,
          },
          center: [50, 50] as [number, number],
          content: 'world 2345',
        },
        children: [
          {
            node: {
              attributes: {
                nodeType: NodeType.IMG,
                style: 'width: 100px; height: 100px;',
                src: 'https://example.com/image.jpg',
                htmlTagName: '<img>',
                ariaLabel: 'image description',
                storyContent:
                  'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
              },
              id: '3',
              indexId: 20,
              rect: {
                x: 0,
                y: 0,
                width: 100,
                height: 100,
                left: 0,
                top: 0,
              },
              center: [50, 50] as [number, number],
              content: '',
            },
            children: [],
          },
          {
            node: {
              attributes: {
                nodeType: NodeType.IMG,
                style: 'width: 100px; height: 100px;',
                src: 'https://example.com/image.jpg',
                htmlTagName: '<img>',
                ariaLabel: 'image description',
                storyContent:
                  'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
              },
              id: '3',
              indexId: 20,
              rect: {
                x: 0,
                y: 0,
                width: 3,
                height: 3,
                left: 0,
                top: 0,
              },
              center: [50, 50] as [number, number],
              content: '',
            },
            children: [],
          },
          {
            node: {
              attributes: {
                nodeType: NodeType.CONTAINER,
              },
              id: '3',
              indexId: 20,
              rect: {
                x: 0,
                y: 0,
                width: 3,
                height: 3,
                left: 0,
                top: 0,
              },
              center: [50, 50] as [number, number],
              content: '',
            },
            children: [
              {
                node: {
                  attributes: {
                    nodeType: NodeType.IMG,
                  },
                  id: '3',
                  indexId: 20,
                  rect: {
                    x: 0,
                    y: 0,
                    width: 3,
                    height: 3,
                    left: 0,
                    top: 0,
                  },
                  center: [50, 50] as [number, number],
                  content: '',
                },
                children: [
                  {
                    node: {
                      attributes: {
                        nodeType: NodeType.IMG,
                        style: 'width: 100px; height: 100px;',
                        src: 'https://example.com/image.jpg',
                        htmlTagName: '<img>',
                        ariaLabel: 'image description',
                        storyContent:
                          'Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every rustling branch, was said to carry the whispers of the ancient spirits who protected the forest. Elara often spent her evenings perched on the boundary stones, listening intently to the murmur of the leaves, wondering what secrets they concealed.',
                      },
                      id: '3222',
                      indexId: 20,
                      rect: {
                        x: 0,
                        y: 0,
                        width: 100,
                        height: 100,
                        left: 0,
                        top: 0,
                      },
                      center: [50, 50] as [number, number],
                      content: 'world 2345',
                    },
                    children: [],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  };

  it('should be able to describe tree', async () => {
    const description = descriptionOfTree(tree);
    // console.log(description);
    expect(description).toMatchSnapshot();
  });

  it('should be able to describe tree, filterNonTextContent = true', async () => {
    const description = descriptionOfTree(tree, 20, true);
    // console.log(description);
    expect(description).toMatchSnapshot();
  });
});
