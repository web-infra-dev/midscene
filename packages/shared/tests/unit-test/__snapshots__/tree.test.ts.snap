// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > should be able to describe tree 1`] = `
"<container id="1" markerId="19"  left="0" top="0" width="100" height="100">
  Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every ...
  <text id="2" markerId="999" ariaLabel="image description, it could be a long text, very loooooooooooooooooooooooooooooooooooooooooong" left="0" top="0" width="100" height="100">
    world
  </text>
  <img id="3" markerId="20" ariaLabel="image description" storyContent="Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every ..." left="0" top="0" width="100" height="100">
    world 2345
    <img id="3" markerId="20" ariaLabel="image description" storyContent="Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every ..." left="0" top="0" width="100" height="100">
    </img>
      <>
        <img id="3222" markerId="20" ariaLabel="image description" storyContent="Legend had it that the Whispering Woods held an ancient secret, one that connected the world of man and magic, of reality and dream. Each leaf, every ..." left="0" top="0" width="100" height="100">
          world 2345
        </img>
      </>
  </img>
</container>"
`;

exports[`utils > should be able to describe tree, filterNonTextContent = true 1`] = `
"<container id="1" markerId="19"  left="0" top="0" width="100" height="100">
  Legend had it that t...
  <text id="2" markerId="999" ariaLabel="image description, i..." left="0" top="0" width="100" height="100">
    world
  </text>
  <img id="3" markerId="20" ariaLabel="image description" storyContent="Legend had it that t..." left="0" top="0" width="100" height="100">
    world 2345
        <img id="3222" markerId="20" ariaLabel="image description" storyContent="Legend had it that t..." left="0" top="0" width="100" height="100">
          world 2345
        </img>
  </img>
</container>"
`;
